lib/CL/clEnqueueFillBuffer.c

```diff
@@ -0,0 +1,113 @@
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in
+   all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+   THE SOFTWARE.
+*/
+
+#include "pocl_util.h"
+#include <string.h>
+
+#ifdef __MSC_VER
+#  include <intrin.h>
+#  define __builtin_popcount __popcnt
+#endif
+
+
+extern CL_API_ENTRY cl_int CL_API_CALL
+POname(clEnqueueFillBuffer)(cl_command_queue  command_queue,
+                           cl_mem            buffer,
+                           const void *      pattern,
+                           size_t            pattern_size,
+                           size_t            offset,
+                           size_t            size,
+                           cl_uint           num_events_in_wait_list,
+                           const cl_event*   event_wait_list,
+                           cl_event*         event)
+CL_API_SUFFIX__VERSION_1_2
+{
+  int errcode = CL_SUCCESS;
+  _cl_command_node *cmd = NULL;
+
+  POCL_RETURN_ERROR_COND((command_queue == NULL), CL_INVALID_COMMAND_QUEUE);
+
+  POCL_RETURN_ERROR_COND((buffer == NULL), CL_INVALID_MEM_OBJECT);
+
+  POCL_RETURN_ERROR_ON((buffer->type != CL_MEM_OBJECT_BUFFER), CL_INVALID_MEM_OBJECT,
+                       "buffer is not a CL_MEM_OBJECT_BUFFER\n");
+
+  POCL_RETURN_ERROR_ON((buffer->flags & CL_MEM_READ_ONLY), CL_INVALID_MEM_OBJECT,
+                       "buffer is CL_MEM_READ_ONLY\n");
+
+  POCL_RETURN_ERROR_ON((command_queue->context != buffer->context), CL_INVALID_CONTEXT,
+                       "buffer and command_queue are not from the same context\n");
+
+  POCL_RETURN_ERROR_COND((event_wait_list == NULL && num_events_in_wait_list > 0),
+                         CL_INVALID_EVENT_WAIT_LIST);
+
+  POCL_RETURN_ERROR_COND((event_wait_list != NULL && num_events_in_wait_list == 0),
+                         CL_INVALID_EVENT_WAIT_LIST);
+
+  errcode = pocl_buffer_boundcheck(buffer, offset, size);
+  if (errcode != CL_SUCCESS)
+    return errcode;
+
+  /* CL_INVALID_VALUE if pattern is NULL or if pattern_size is 0
+   * or if pattern_size is not one of {1, 2, 4, 8, 16, 32, 64, 128}. */
+  POCL_RETURN_ERROR_COND((pattern == NULL), CL_INVALID_VALUE);
+  POCL_RETURN_ERROR_COND((pattern_size == 0), CL_INVALID_VALUE);
+  POCL_RETURN_ERROR_COND((pattern_size > 128), CL_INVALID_VALUE);
+
+  POCL_RETURN_ERROR_ON((__builtin_popcount(pattern_size) > 1), CL_INVALID_VALUE,
```

Comment:
OK, i've checked that both clang & gcc supports it, also added an MSVC macro. Is that good enough ?